File: backend/src/main/scala/domains/usergroup/objects/response/UserGroupDetail.scala
package domains.usergroup.objects.response

import domains.usergroup.objects.*

import domains.user.objects.Username
import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.{deriveDecoder, deriveEncoder}

import java.time.Instant
import scala.util.Try

/** 用户组详情响应，包含基础资料、所有者、成员列表和审计时间。 */
final case class UserGroupDetail(
  id: UserGroupId,
  slug: UserGroupSlug,
  name: UserGroupName,
  description: UserGroupDescription,
  ownerUsername: Username,
  members: List[UserGroupMember],
  createdAt: Instant,
  updatedAt: Instant
)

/** 提供用户组详情响应编解码和从内部 UserGroup 转换的入口。 */
object UserGroupDetail:
  private given Encoder[Instant] = Encoder.encodeString.contramap(_.toString)
  private given Decoder[Instant] = Decoder.decodeString.emap { value =>
    Try(Instant.parse(value)).toEither.left.map(_.getMessage)
  }

  given Encoder[UserGroupDetail] = deriveEncoder[UserGroupDetail]
  given Decoder[UserGroupDetail] = deriveDecoder[UserGroupDetail]

  /** 将内部用户组模型转换为 API 详情响应，不改变成员排序。 */
  def fromUserGroup(group: UserGroup): UserGroupDetail =
    UserGroupDetail(
      id = group.id,
      slug = group.slug,
      name = group.name,
      description = group.description,
      ownerUsername = group.ownerUsername,
      members = group.members,
      createdAt = group.createdAt,
      updatedAt = group.updatedAt
    )

File: frontend/src/objects/usergroup/response/UserGroupDetail.ts
import type { Username } from '@/objects/user/Username'
import { fromUsernameContract } from '@/objects/user/Username'
import type { UserGroupDescription } from '@/objects/usergroup/UserGroupDescription'
import { fromUserGroupDescriptionContract } from '@/objects/usergroup/UserGroupDescription'
import type { UserGroupId } from '@/objects/usergroup/UserGroupId'
import { fromUserGroupIdContract } from '@/objects/usergroup/UserGroupId'
import type { UserGroupMember } from '@/objects/usergroup/UserGroupMember'
import { fromUserGroupMemberContract } from '@/objects/usergroup/UserGroupMember'
import type { UserGroupName } from '@/objects/usergroup/UserGroupName'
import { fromUserGroupNameContract } from '@/objects/usergroup/UserGroupName'
import type { UserGroupSlug } from '@/objects/usergroup/UserGroupSlug'
import { fromUserGroupSlugContract } from '@/objects/usergroup/UserGroupSlug'
import type { AuditFields } from '@/objects/shared/AuditFields'
import { fromAuditFieldsContract } from '@/objects/shared/AuditFields'
import { readArray, readRecord, readString } from '@/objects/shared/PageResponse'

export type UserGroupDetail = AuditFields & {
  id: UserGroupId
  slug: UserGroupSlug
  name: UserGroupName
  description: UserGroupDescription
  ownerUsername: Username
  members: UserGroupMember[]
}

export function fromUserGroupDetailContract(value: unknown, label = 'user group detail'): UserGroupDetail {
  const group = readRecord(value, label)
  return {
    ...fromAuditFieldsContract(value, label),
    id: fromUserGroupIdContract(readString(group.id, `${label} id`), `${label} id`),
    slug: fromUserGroupSlugContract(readString(group.slug, `${label} slug`), `${label} slug`),
    name: fromUserGroupNameContract(readString(group.name, `${label} name`), `${label} name`),
    description: fromUserGroupDescriptionContract(readString(group.description, `${label} description`), `${label} description`),
    ownerUsername: fromUsernameContract(readString(group.ownerUsername, `${label} owner username`), `${label} owner username`),
    members: readArray(group.members, `${label} members`, fromUserGroupMemberContract),
  }
}
